### PR TITLE
NF: adds updated custom dataset+file extractors

### PR DESCRIPTION
**Pull request by jsheunis**
This PR addresses #328. It:
- renames extractor from `custom` to `generic json` (can we get consensus on this?)
- adds the genericjson_dataset` and `genericjson_file` modules (split from `custom.py`, and upgrading to the latest generation of extractor base classes)
- adds/updates relevant tests
- replaces metadata source file specification via dataset config with specifying during the call via extraction arguments
- updates extractor docstrings to provide more useful information on usage

TODO:
- similarly update other metalad-shipped extractors: to be done in separate PRs

### CODE REVIEW

**Comment by jsheunis** [comment]
I'll write up my challenges with implementing result rendering here. I think the problem I'm facing is manifold:

#### 1 - I don't know when it is more or less appropriate to yield a datalad results record vs raising an exception

To give a specific example related to this PR, the extractor class has an instance method `get_required_content()` (which overwrites the base class method) that should contain code to ensure that all necessary file content is locally available in order to continue with the extraction. If a required file is not available, I am not sure whether this should yield an impossible/error result, vs raise exception. Also, further down the line during the `extract()` method call, where the actual required file is being opened and read from, should I at that point just expect it to be there and let standard error handling take over? Or should it also yield a results record?

#### 2 - I am unsure what to do when a specific function should return a value (e.g. a `dict`), but the logic inside the function might require an exception to be raised or a results record to be yielded.

How should I handle the returned value? Should it always be a results record that is unpacked when returned, and if it's an `ok` result then the associated value is used. What is a good, consistent way to do the unpacking? To continue from the example above, in the `extract()` method there is an  `immediate_data` variable that should receive a `dict`, which is returned from the part of the code that does the actual extraction. But during the actual extraction, some things can go wrong (e.g. a `FileNotFoundError`). So how do I create a seamless process that either returns the actual `dict`, or the error, and importantly: how is that returned `dict` or results record unpacked so that theres no mismatch between what is expected and what is actually returned?

#### Next steps

It could very well be that I just have an insufficient understanding of generators and results rendering etc, but I would personally benefit from more documentation on how the results rendering of datalad and its extensions is intended to be used. Everything I know about it has been pieced together from looking at existing code, which could either be good or bad examples (and I cannot comfortably say which is which).
-

**Comment by jsheunis** [comment]
I'm busy upgrading the custom file-level extractor to the newest gen and encountering a challenge that I'd like some input on.

**The new:** The latest [`FileMetadataExtractor`](https://github.com/datalad/datalad-metalad/blob/master/datalad_metalad/extractors/base.py#L176) operates on a single file. An extractor that inherits from this class can specify whether the file content is required locally for extraction, by returning a `bool` type via the `is_content_required()` method.

**The old:** The so-called legacy custom extractor operates on files with json content that have been added to a dataset specifically for the purpose of describing a file in the dataset. These json files can be stored either in `.metadata/content/*` (e.g. `.metadata/content/myfile.json`)or as a sidecar file alongside the actual file (e.g. `mydir/_metadata_myfile.json` for `mydir/myfile.png`).

**So the question is:** if we think about how a user will intuitively interact with this updated extraction process, would they:
1. specify the actual filename (i.e. `mydir/myfile.png`), and then the extractor (together with some configuration) will determined which json file it should extract the information from (e.g. `mydir/_metadata_myfile.json`)? In this case, the `is_content_required()` should return `False` because the actual file is not needed, but then another method would be needed to get the actual json files that contain the to-be-extracted metadata. As is, the easiest (only?) place to do this would probably be inside the `extract()` method. 
2. specify the json filename containing the to-be-extracted metadata (e.g. `mydir/_metadata_myfile.json` or `.metadata/content/myfile.json`). This would not require an additional method to get the file content, as it would be handled by the existing `is_content_required()`, but IMO it's less intuitive to have to specify the json file (wherever it is located).

**Comment by jsheunis** [comment]
I went with this option:

> 1. specify the actual filename (i.e. mydir/myfile.png), and then the extractor (together with some configuration) will determined which json file it should extract the information from (e.g. mydir/_metadata_myfile.json)? In this case, the is_content_required() should return False because the actual file is not needed, but then another method would be needed to get the actual json files that contain the to-be-extracted metadata. As is, the easiest (only?) place to do this would probably be inside the extract() method.

**Comment by christian-monch** [comment]
> #### 1 - I don't know when it is more or less appropriate to yield a datalad results record vs raising an exception
> To give a specific example related to this PR, the extractor class has an instance method `get_required_content()` (which overwrites the base class method) that should contain code to ensure that all necessary file content is locally available in order to continue with the extraction. If a required file is not available, I am not sure whether this should yield an impossible/error result, vs raise exception. [...]

I think it should be an "impossible"-result, since the precondition for extraction can not be met.

> [...] Also, further down the line during the `extract()` method call, where the actual required file is being opened and read from, should I at that point just expect it to be there and let standard error handling take over? Or should it also yield a results record?

Since ``get_required_content()`` should have retrieved the content, this situation would constitute an error. So, using standard error handling seems a natural choice.

**Comment by jsheunis** [comment]
Discussed in a meeting:
- File-level `get_required_content` should rather be an implementation in metalad
- Naming: “custom” sounds misleading (it describes the metadata, rather than the extractor); consider renaming, perhaps “generic_json”?
- Move specification of metadata source locations to metalad call parameter (more standard wrt metalad, and captures provenance of the call)
- Consider changing namespace for default metadata sources: look at `.datalad` rather than `.metadata`, so as not to pollute the existing namespace
- Investigate whether BIDS file extractor can derive from this "generic_json" extractor, given the sidecar functionality (and what about dataset-level?)

Also, found this existing issue about `get_required_content` at the file-level: https://github.com/datalad/datalad-metalad/issues/287

**Comment by christian-monch** [comment]
> [...]
> * Consider changing namespace for default metadata sources: look at `.datalad` rather than `.metadata`, so as not to pollute the existing namespace
> [...]

On a second thought I would really discourage to let datalad/metalad users write into ``.datalad``. We should reserve this namesspace for us. There is also a very real collision risk, because the old metalad uses ``.datalad/metadata`` to store metadata. This comes in addition to the risk of "destroying" a dataset by unintentionally deleting or modifying content in ``.datalad``.

**Comment by jsheunis** [comment]
> On a second thought I would really discourage to let datalad/metalad users write into .datalad. We should reserve this namesspace for us. There is also a very real collision risk, because the old metalad uses .datalad/metadata to store metadata. This comes in addition to the risk of "destroying" a dataset by unintentionally deleting or modifying content in .datalad.

Good point, I didn't think of this. Does anyone feel strongly about any alternative to `<dataset_path>/.metadata` for all metadata added by users for the purpose of `generic_json` extraction? If not, I'll stick to that.

**Review by christian-monch — APPROVED** [review]
Thx @jsheunis. Looks good to me.

**Comment by jsheunis** [comment]
Current TODOs:

- [ ] Naming: “custom” sounds misleading (it describes the metadata, rather than the extractor); consider renaming, perhaps “generic_json”?
- [ ] Move specification of metadata source locations to metalad call parameter (more standard wrt metalad, and captures provenance of the call)

File: datalad_metalad/extract.py [hunk 1]
```diff
@@ -349,116 +353,83 @@ def custom_result_renderer(res, **kwargs):
...
-            "performing legacy file level metadata "
-            "extraction for file at %s/%s",
-            ep.source_dataset.path,
-            ep.file_tree_path)
-
-        yield from legacy_extract_file(ep)
+            "performing legacy %s-level metadata "
+            "extraction for %s at %s",
+            extractor_type,
+            extractor_type,
```
Comment (christian-monch) [inline datalad_metalad/extract.py:369]:
```suggestion
            ep.extractor_name,
```
Should that be ``extractor_name``?

File: datalad_metalad/extract.py [hunk 2]
```diff
@@ -349,116 +353,83 @@ def custom_result_renderer(res, **kwargs):
...
-        "performing file level extracting for file at %s/%s",
-        ep.source_dataset.path,
-        ep.file_tree_path)
-
-    file_info = get_file_info(ep.source_dataset, ep.file_tree_path)
-    extractor = ep.extractor_class(
+            "performing %s-level metadata "
+            "extraction for %s at %s",
+            extractor_type,
+            extractor_type,
```
Comment (christian-monch) [inline datalad_metalad/extract.py:394]:
```suggestion
            ep.extractor_name,
```

File: datalad_metalad/extractors/tests/test_genericjson_file.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    #     path=str(ds.pathobj / 'sub' / 'one'),
+    #     type='file',
+    #     # specific message does vary a lot across platforms
+    #     #message=
+    #     action="meta_extract",
+    #     status="error",
+    #     # exception=ce,
+    #     # message=("cannot load data from target path: %s", ce),
+    #     metadata={},
+    # )
```
Comment (christian-monch) [inline datalad_metalad/extractors/tests/test_genericjson_file.py:92]:
```suggestion
```
Let's not leave commented code in the code-base. :-)

**Review by christian-monch — APPROVED** [review]
Nice, thx a lot @jsheunis 

Just a few comments. @jsheunis: feel free to merge after addressing those

File: datalad_metalad/extractors/tests/test_genericjson_file.py (hunk 3 above)
Comment (jsheunis) [reply datalad_metalad/extractors/tests/test_genericjson_file.py:92]: missed that, thanks!

**Pull request merged**